File: runner.js
// runner.js - Бесконечный бегун на JavaScript (Node.js)
const fs = require('fs');
const keypress = require('keypress');
const readline = require('readline');

const WIDTH = 40;
const HEIGHT = 10;
const GROUND = HEIGHT - 1;
const PLAYER = '@';
const OBSTACLE = '#';
const RECORD_FILE = 'runner_record.json';

class RunnerGame {
    constructor() {
        this.width = WIDTH;
        this.height = HEIGHT;
        this.ground = GROUND;
        this.playerX = 5;
        this.playerY = GROUND;
        this.playerVy = 0;
        this.onGround = true;
        this.obstacles = [];
        this.score = 0;
        this.speed = 1;
        this.gameOver = false;
        this.paused = false;
        this.running = true;
        this.record = this.loadRecord();
        this.lastUpdate = Date.now();
        this.timer = null;
    }

    loadRecord() {
        try {
            const data = fs.readFileSync(RECORD_FILE, 'utf8');
            return JSON.parse(data).record || 0;
        } catch { return 0; }
    }

    saveRecord() {
        fs.writeFileSync(RECORD_FILE, JSON.stringify({ record: this.record }));
    }

    jump() {
        if (this.onGround && !this.gameOver && !this.paused) {
            this.playerVy = -3.0;
            this.onGround = false;
        }
    }

    spawnObstacle() {
        this.obstacles.push(this.width - 1);
    }

    updatePhysics() {
        this.playerVy += 0.2;
        this.playerY += this.playerVy;
        if (this.playerY >= this.ground) {
            this.playerY = this.ground;
            this.playerVy = 0;
            this.onGround = true;
        }
        if (this.playerY < 0) {
            this.playerY = 0;
            this.playerVy = 0;
        }
    }

    update() {
        if (this.gameOver || this.paused) return;

        // Движение препятствий
        for (let i = this.obstacles.length - 1; i >= 0; i--) {
            this.obstacles[i]--;
            if (this.obstacles[i] < 0) this.obstacles.splice(i, 1);
        }

        // Столкновение
        const playerFloor = Math.floor(this.playerY);
        for (const x of this.obstacles) {
            if (x === this.playerX && (playerFloor >= this.ground - 1 || playerFloor === this.ground - 1)) {
                this.gameOver = true;
                if (this.score > this.record) {
                    this.record = this.score;
                    this.saveRecord();
                }
                return;
            }
        }

        this.score++;
        this.speed = 1 + Math.floor(this.score / 10);

        if (Math.random() < 0.2 + Math.min(0.4, this.score / 200)) {
            this.spawnObstacle();
        }

        this.updatePhysics();
    }

    draw() {
        console.clear();
        console.log('═'.repeat(this.width));
        console.log(`  Счёт: ${this.score}   Рекорд: ${this.record}   Скорость: ${this.speed}`);
        console.log('═'.repeat(this.width));

        const grid = Array.from({ length: this.height }, () => Array(this.width).fill(' '));
        // Препятствия
        for (const x of this.obstacles) {
            if (x >= 0 && x < this.width) {
                grid[this.ground][x] = OBSTACLE;
            }
        }
        // Игрок
        const y = Math.floor(this.playerY);
        if (y >= 0 && y < this.height) {
            grid[y][this.playerX] = PLAYER;
        }

        for (const row of grid) {
            console.log('│' + row.join('') + '│');
        }
        console.log('═'.repeat(this.width));
        const status = this.paused ? 'ПАУЗА' : 'ИГРА';
        console.log(`  ${status}  |  Пробел/↑ - прыжок  |  P - пауза  |  R - рестарт  |  Q - выход`);
    }

    reset() {
        this.playerY = this.ground;
        this.playerVy = 0;
        this.onGround = true;
        this.obstacles = [];
        this.score = 0;
        this.speed = 1;
        this.gameOver = false;
        this.paused = false;
    }

    handleKey(ch, key) {
        if (!key) return;
        if (key.name === 'space' || key.name === 'up') {
            this.jump();
        } else if (key.name === 'p') {
            this.paused = !this.paused;
        } else if (key.name === 'r') {
            this.reset();
        } else if (key.name === 'q') {
            this.running = false;
            process.stdin.pause();
            process.exit(0);
        }
    }

    run() {
        // Настройка keypress
        keypress(process.stdin);
        process.stdin.setRawMode(true);
        process.stdin.resume();
        process.stdin.on('keypress', (ch, key) => this.handleKey(ch, key));

        const gameLoop = () => {
            if (!this.running) return;
            const now = Date.now();
            if (now - this.lastUpdate >= 1000 / 60) {
                this.update();
                this.draw();
                this.lastUpdate = now;
            }
            this.timer = setTimeout(gameLoop, 16);
        };
        gameLoop();
    }
}

const game = new RunnerGame();
game.run();

File: runner.py

---

## 💻 Код на 7 языках

### 1. Python – `runner.py`

```python
#!/usr/bin/env python3
# runner.py - Бесконечный бегун на Python

import os
import sys
import time
import random
import json
import threading
from dataclasses import dataclass
from typing import List, Optional

try:
    import keyboard
    from colorama import init, Fore, Style
    init(autoreset=True)
except ImportError:
    print("Установите зависимости: pip install keyboard colorama")
    sys.exit(1)

# Константы
WIDTH = 40
HEIGHT = 10
GROUND = HEIGHT - 1
PLAYER = '@'
OBSTACLE = '#'
EMPTY = ' '
RECORD_FILE = 'runner_record.json'

class RunnerGame:
    def __init__(self):
        self.width = WIDTH
        self.height = HEIGHT
        self.ground = GROUND
        self.player_x = 5
        self.player_y = GROUND
        self.player_vy = 0       # вертикальная скорость
        self.on_ground = True
        self.obstacles: List[int] = []  # x-координаты препятствий
        self.score = 0
        self.speed = 1
        self.game_over = False
        self.paused = False
        self.running = True
        self.record = self.load_record()
        self.lock = threading.Lock()

    def load_record(self) -> int:
        try:
            with open(RECORD_FILE, 'r') as f:
                data = json.load(f)
                return data.get('record', 0)
        except:
            return 0

    def save_record(self):
        with open(RECORD_FILE, 'w') as f:
            json.dump({'record': self.record}, f)

    def jump(self):
        if self.on_ground and not self.game_over and not self.paused:
            self.player_vy = -3.0
            self.on_ground = False

    def spawn_obstacle(self):
        # Препятствия появляются справа
        self.obstacles.append(self.width - 1)

    def update_physics(self):
        # Гравитация
        self.player_vy += 0.2
        self.player_y += self.player_vy

        # Проверка земли
        if self.player_y >= self.ground:
            self.player_y = self.ground
            self.player_vy = 0
            self.on_ground = True

        # Не выше потолка
        if self.player_y < 0:
            self.player_y = 0
            self.player_vy = 0

    def update(self):
        if self.game_over or self.paused:
            return

        # Движение препятствий
        for i in range(len(self.obstacles)):
            self.obstacles[i] -= 1
        # Удалить ушедшие за левую границу
        self.obstacles = [x for x in self.obstacles if x >= 0]

        # Столкновение (если препятствие на земле или выше, а игрок не в воздухе)
        for x in self.obstacles:
            # Проверяем, совпадает ли позиция препятствия с позицией игрока
            if x == self.player_x and self.player_y >= self.ground - 1:  # игрок на земле
                self.game_over = True
                if self.score > self.record:
                    self.record = self.score
                    self.save_record()
                return
            # Если препятствие на уровне выше, но игрок не допрыгнул
            # Для простоты проверяем совпадение по x и что игрок не на земле
            if x == self.player_x and self.player_y == self.ground - 1:
                self.game_over = True
                if self.score > self.record:
                    self.record = self.score
                    self.save_record()
                return

        # Счёт
        self.score += 1
        self.speed = 1 + self.score // 10

        # Спавн новых препятствий
        if random.random() < 0.2 + min(0.4, self.score / 200):
            self.spawn_obstacle()

        # Обновление физики игрока
        self.update_physics()

    def draw(self):
        os.system('cls' if os.name == 'nt' else 'clear')
        print('═' * self.width)
        print(f'  Счёт: {self.score}   Рекорд: {self.record}   Скорость: {self.speed}')
        print('═' * self.width)

        # Создаём сетку
        grid = [[' ' for _ in range(self.width)] for _ in range(self.height)]
        # Размещаем препятствия
        for x in self.obstacles:
            if 0 <= x < self.width:
                grid[self.ground][x] = Fore.RED + OBSTACLE + Style.RESET_ALL
        # Размещаем игрока
        if 0 <= self.player_y < self.height:
            grid[int(self.player_y)][self.player_x] = Fore.GREEN + PLAYER + Style.RESET_ALL

        # Вывод
        for row in grid:
            print('│' + ''.join(row) + '│')
        print('═' * self.width)
        status = "ПАУЗА" if self.paused else "ИГРА"
        print(f'  {status}  |  Пробел/↑ - прыжок  |  P - пауза  |  R - рестарт  |  Q - выход')

    def handle_input(self):
        while self.running:
            try:
                if keyboard.is_pressed('space') or keyboard.is_pressed('up'):
                    with self.lock:
                        self.jump()
                elif keyboard.is_pressed('p'):
                    with self.lock:
                        self.paused = not self.paused
                    time.sleep(0.2)
                elif keyboard.is_pressed('r'):
                    with self.lock:
                        self.reset()
                    time.sleep(0.2)
                elif keyboard.is_pressed('q'):
                    self.running = False
                    break
            except:
                pass
            time.sleep(0.05)

    def reset(self):
        self.player_y = self.ground
        self.player_vy = 0
        self.on_ground = True
        self.obstacles = []
        self.score = 0
        self.speed = 1
        self.game_over = False
        self.paused = False

    def run(self):
        # Запуск потока для ввода
        input_thread = threading.Thread(target=self.handle_input, daemon=True)
        input_thread.start()

        # Основной игровой цикл (60 FPS)
        last_update = time.time()
        while self.running:
            now = time.time()
            if now - last_update >= 1.0 / 60:
                with self.lock:
                    self.update()
                    self.draw()
                last_update = now
            time.sleep(0.01)

if __name__ == "__main__":
    game = RunnerGame()
    try:
        game.run()
    except KeyboardInterrupt:
        print("\nВыход...")
